### DISCUSSION TITLE
Invalid TIFF directory,; tags are not sorted in ascending order

### DISCUSSION THREAD

**Discussion by richirasega** [ImageMagick/ImageMagick#2661] [Help]
Hello,
my OS is Windows Server 2016 standard 64Bit, I've updated Imagemagick by a previous version to the latest (7.0.10-Q16-HDRI) and now I'm receiveing this warning while converting EPS to JPG:
**Invalid TIFF directory,; tags are not sorted in ascending order**

The JPG created is now wrong, maybe also due the warning.
This is what I run by command line:
`convert.exe -strip -format "jpg" -density 256X256  -depth 8 -units PixelsPerInch -colorspace RGB  SOURCEFILE.eps -resize 300x300 -background white -flatten DESTFILE.jpg`

Could anyone suggest me a fix or what to check ?

Thank you

**Comment by dlemstra** [comment]
Can you attach the file? Without it we have to make guesses.

**Comment by richirasega** [comment]
Sure, you could check it by here (this is just an example, but the results is always the same!)
https://drive.google.com/file/d/1yeztVu45Etxcprvwhq7w0xJGbozkGTWv/view?usp=sharing

**Comment by t246246** [comment]
According to the tiff spec, The entries in an IFD must be sorted in ascending order by Tag. https://www.awaresystems.be/imaging/tiff/specification/TIFF6.pdf
So this should be a message when reading invalid tiff that would be created temporarily during conversion.

As I cannot access your google drive to get files, I cannot confirm it.  Also as your input is eps, ghostscript is used to read it.  So need your ghostscript version too.

**Reply by richirasega** [reply]
I've released the access to the file. 
Anyway this test file was already converted using this exact code with a previous version of ImageMagick and GS, maybe something changed in new versions ...

**Reply by t246246** [reply]
Your file is mixture of PostScript and tiff, I did not know eps accept such thing.

$ file TEST_RIC.eps
TEST_RIC.eps: DOS EPS Binary File Postscript starts at byte 32 length 1116843 TIFF starts at byte 1116875 length 8011

So I extract tiff part and check it and try to convert it to jpg,

$ dd bs=1 skip=1116875 count=8011 if=TEST_RIC.eps of=temp.tif
$ tiffinfo.exe temp.tif
TIFF Directory at offset 0x1ea9 (7849)
  Subfile Type: (0 = 0x0)
  Image Width: 45 Image Length: 45
  Bits/Sample: 8
  Compression Scheme: None
  Photometric Interpretation: palette color (RGB from colormap)
  Extra Samples: 1<assoc-alpha>
  Samples/Pixel: 2
  Planar Configuration: single image plane
  Color Map: (present)
  Photoshop Data: <present>, 2244 bytes
TIFFReadDirectoryCheckOrder: Warning, Invalid TIFF directory; tags are not sorted in ascending order.

$ magick temp.tif temp.jpg
magick.exe: Invalid TIFF directory; tags are not sorted in ascending order. `TIFFReadDirectoryCheckOrder' @ warning/tiff.c/TIFFWarnings/1037.

Yes, it warns, but it is just a warning.  So if you get the desired jpg files, just ignore.
add -quiet to your command.

**Reply by richirasega** [reply]
No the file created is not correct, now I've switched to another solution and convert by Ghostscript to JPG and then resize the JPG itself and now it works

**Reply by t246246** [reply]
How wrong the jpg file? What is the previous version of ImageMagick and GS?

If you omit -flattern, you get 2 files, DESTFILE-0.jpg and DESFILE-1.jpg.  Any of which is correct?

**Reply by richirasega** [reply]
I don't have the previous version number.
Yes, I get 2 files and one of those are correct, the second one is always a blue square that I can't figure out where it comes from

**Reply by t246246** [reply]
I tested slightly older version 7.0.10-11 Q16 on MSYS2, first one is converted from PostScript, so it is clear; wheres 2nd one is from tiff, so it is jaggy. (2nd one is created from ImageMagick itself, not using gs.)

But I did not got the blue square. So it must be a bug.

**Comment by t246246** [comment]
I installed the newest one and gs 9.53 on Windows 10 and confirmed this odd behavior.

% magick -version
Version: ImageMagick 7.0.10-31 Q16 x64 2020-09-27 http://www.imagemagick.org
Copyright: Copyright (C) 1999-2018 ImageMagick Studio LLC
License: http://www.imagemagick.org/script/license.php
Visual C++: 192729111
Features: Cipher DPC HDRI Modules OpenCL OpenMP(2.0)
Delegates (built-in): bzlib cairo flif freetype gslib heic jng jp2 jpeg lcms lqr lzma openexr pangocairo png ps raw rsvg tiff webp xml zlib


You already have another solution, but currently it works (as you need only 1st image of eps).

% magick TEST_RIC.eps[0] -verbose -strip -format "jpg" -density 256X256 -depth 8 -units PixelsPerInch -colorspace RGB -resize 300x300 -background white DESTFILE.jpg

However, 

% magick TEST_RIC.eps[1] -verbose -strip -format "jpg" -density 256X256 -depth 8 -units PixelsPerInch -colorspace RGB -resize 300x300 -background white DESTFILE.jpg

this one becomes blue (rgb(0,0,50)) square.

Try to analyze it step by step:

> magick -quiet TEST_RIC.eps[0] DEST0.tif
> magick -quiet TEST_RIC.eps[1] DEST1.tif

> magick identify DEST0.tif DEST1.tif
DEST0.tif TIFF 45x45 45x45+0+0 16-bit CMYK 64251B 0.000u 0:00.001
DEST1.tif TIFF 45x45 45x45+0+0 8-bit Grayscale Gray 6576B 0.000u 0:00.000

> magick DEST0.tif -depth 8 -colorspace RGB DEST1.tif -flatten DEST.tif

This DEST.tif looks good.  (For now I ignore density/size, I mean it is at least not a blue square.)

First problem is:

When -colorspace RGB is applied to TEST_RIC.eps[1], it becomes square image.
> magick TEST_RIC.eps[1] -colorspace RGB x.tif

checking histogram, 
c:\work>magick -quiet TEST_RIC.eps[1] -format %c histogram:info:
    180: (0,0,0,0) #00000000 none
    67: (0,0,0,255) #000000FF black
    8: (51,51,51,255) #333333FF grey20
    99: (102,102,102,255) #666666FF grey40
    99: (153,153,153,255) #999999FF grey60
    3: (204,204,204,255) #CCCCCCFF grey80
    1569: (255,255,255,255) #FFFFFFFF white

c:\work>magick -quiet TEST_RIC.eps[1] -colorspace RGB -format %c histogram:info:
    2025: (0,0,8.44172,255) #000008FF rgba(0%,0%,3.31048%,1)

c:\work>magick -quiet TEST_RIC.eps[1] -colorspace sRGB -format %c histogram:info:
    180: (0,0,0,0) #00000000 none
    67: (0,0,0,255) #000000FF black
    8: (51,51,51,255) #333333FF grey20
    99: (102,102,102,255) #666666FF grey40
    99: (153,153,153,255) #999999FF grey60
    3: (204,204,204,255) #CCCCCCFF grey80
    1569: (255,255,255,255) #FFFFFFFF white

So you should use sRGB instead of RGB in these days.

There is another problem. Look carefully DEST1.tif.  It's drawing is slightly shift down.
So the result of -flatten is wrong.  I found no way to solve it.

**Comment by urban-warrior** [comment]
Your image is a EPT image which includes two image blobs, one EPS, and the other TIFF.  EPS is interpreted by Ghostscript.  The TIFF blob is interpreted by the libtiff delegate library.  EPS may change depending on your release of Ghostscript.  The TIFF blob appears to violate the TIFF standard of sorted directories.  In addition, sometimes Adobe has artifacts that are proprietary that are not properly interpreted by the TIFF delegate library or Ghostscript.

Ghostscript results are driven by the selected driver.    Since the EPS blob is CMYK, we use pamcmyk.  If you instead specify -colorspace sRGB, we use the pngalpha driver and instead return a sRGB image with an alpha channel.

Previous versions of ImageMagick would prefer the TIFF blob unless it was corrupt and then it would instead return the EPS blob.  Recent releases return both blobs, EPS & TIFF and lets the user select which one they prefer (e.g. TEST_RIC.eps[1]).

As you can see, there are plenty of opportunity for variability and opportunity to return differing results depending on ImageMagick, Ghostscript, libtiff, colorspace selection, etc.  This differs from a format like PNG which will always return the same consistent results.
